In [1]:
!git clone https://github.com/clovaai/deep-text-recognition-benchmark.git

Cloning into 'deep-text-recognition-benchmark'...
remote: Enumerating objects: 499, done.
remote: Counting objects: 100% (4/4), done.
remote: Compressing objects: 100% (4/4), done.
remote: Total 499 (delta 0), reused 1 (delta 0), pack-reused 495
Receiving objects: 100% (499/499), 3.07 MiB | 23.30 MiB/s, done.
Resolving deltas: 100% (301/301), done.


In [2]:
%cd deep-text-recognition-benchmark

/content/deep-text-recognition-benchmark


# Run sample demo :

download pretrained weights : https://drive.google.com/drive/folders/15WPsuPJDCzhp2SvYZLRj8mAlT3zmoAMW

download from google drive , using gdown


In [3]:
!pip install gdown
import gdown

!gdown 1b59rXuGGmKne1AuHnkgDzoYgKeETNMv9  # file ID

Downloading...
From: https://drive.google.com/uc?id=1b59rXuGGmKne1AuHnkgDzoYgKeETNMv9
To: /content/deep-text-recognition-benchmark/TPS-ResNet-BiLSTM-Attn.pth
100% 199M/199M [00:01<00:00, 194MB/s]


In [4]:
!pip install lmdb

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 299.2/299.2 kB 4.4 MB/s eta 0:00:00


In [5]:
!python3 demo.py \
--Transformation TPS --FeatureExtraction ResNet --SequenceModeling BiLSTM --Prediction Attn \
--image_folder demo_image/ \
--saved_model TPS-ResNet-BiLSTM-Attn.pth

model input parameters 32 100 20 1 512 256 38 25 TPS ResNet BiLSTM Attn
loading pretrained model from TPS-ResNet-BiLSTM-Attn.pth
/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py:557: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  warnings.warn(_create_warning_msg(
--------------------------------------------------------------------------------
image_path               	predicted_labels         	confidence score
--------------------------------------------------------------------------------
demo_image/demo_1.png    	available                	0.9999
demo_image/demo_2.jpg    	shakeshack               	0.9530
demo_image/demo_3.png    	london               

rooye 10 ta tasvire demo , dar file demo.py khorooji gereftan

# train DTRB ON PERSIAN License Plate :

mikhaim az transfer learning estefade konim , va shabake RESNET dar bala ro ke ghablan train shode , rooye dataset khodemoon FINE TUNE konim

train our own non-latin dataset :

### WE NEED TO CREATE DATASET WITH lmdb FORMAT


data <br/>
├── gt.txt <br/>
└── test <br/>
    ├── word_1.png <br/>
    ├── word_2.png <br/>
    ├── word_3.png <br/>
    └── ... <br/>

<br/>
 image address  --------    its text <br/>
test/word_1.png -----  Tiredness <br/>
test/word_2.png -----  kills <br/>
test/word_3.png -----  A <br/>
... <br/>

# here , **LABELS ARE TEXTS IN IMAGES**

یک فایل دات تی اکس تی میسازیم <br/>
آدرس هر عکس با متن توش رو قرار میدیم<br/>
یعنی یک پوشه دیتا پر از عکس به علاوه یک فایل جی تی تکست که لیبل ها توشه
این ساختار رو میدیم به فایل <br/> creat_lmbd_dataset.py
<br/>
بعد این میاد از روی عکس ها یک فایل ال ام دی بی میسازه
<br/>
اون فایل ال ام دی بی رو میدیم به train.py
<br/>
روش ترین بشه


میایم توی فولدری که ساخته شده به نام دیپ تکست ریکاگنیشن یک فولدر ایجاد می کنیم به نام دیتا
<br/>

توی این فولدر یک فایل gt.txt  میسازیم
<br/>

و بعد یک فولدر که عکس ها رو بزاریم توش
به نام ترین
<br/>

یک فولدر دیگه هم به نام ولیدیشین ایجاد میکنیم

<br/>
پس حالا چون دو تا فولدر داریم باید دو تا فایل جی تی هم داشته باشیم برای هرکدوم
gt_train.txt
gt_validation.txt

### نکته : بین اسم فایل و متن مربوطه اش نباید اسپیس باشه بلکه باید یک تب بزنیم

<br/>


حالا میریم توی وی اس کد و اونجا فایل تکست رو میسازیم
آدرس عکس + متن تصویر
<br/>

## Labeling :
train/46469fhdhsj.jpg  13M89422
train/834uy884848.jpg  74K48495 <br/>
. <br/>
. <br/>
. <br/>
<br/>
in validation txt file :<br/>
validtaion/h83y3rh7r.jpg  23Y94355



حالا باید کتابخونه فایر رو نصب کنیم :

In [6]:
!pip install fire

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 88.3/88.3 kB 2.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for fire: filename=fire-0.5.0-py2.py3-none-any.whl size=116933 sha256=6ff25fc91391e3a6d281179d52e18c6587d5c6aca5d49e525c1a7d00c732853d
  Stored in directory: /root/.cache/pip/wheels/90/d4/f7/9404e5db0116bd4d43e5666eaa3e70ab53723e1e3ea40c9a95
Successfully built fire


حالا گفته فایل create_lmdb_dataset.py <br/>
رو اجرا بکن

یک پوشه دیتاست ایجادد می کنیم و بعد داخلش یک فولدر ترین ایجاد می کنیم تا نتایج کد پایین رو اون تو ذخیره کنه
<br/>
deep-text-recognition-...  --->  dataset ---> train


In [15]:
!python3 create_lmdb_dataset.py --inputPath data/ --gtFile data/gt_train.txt --outputPath dataset/train

Created dataset with 72 samples


اینجا یک دیتاست ساخته شد

از روی عکس ها و لیبل ها اومد و فایل ام دی بی ساخت در آدرس زیر ⁉ <br/>

dataset --> train  ----> data.mdb

# az file haye .mdb dar TRAIN estefade mikone

حالا همین مسیر رو برای ولیدیشن انجام میدیم:
<br/>
و میگیم در این مسیر ذخیره کن نتایج و فایل های ام دی بی رو
<br/>
deep-text-recognition... / dataset / validation

In [ ]:
!python3 create_lmdb_dataset.py --inputPath data/ --gtFile data/gt_validation.txt --outputPath dataset/validation

# TRAIN :

In [18]:
!python3 train.py \
--train_data dataset/train --valid_data dataset/validation \
--select_data / --batch_ratio 1 --batch_max_length 8 --valInterval 100 \
--Transformation TPS --FeatureExtraction ResNet --SequenceModeling BiLSTM --Prediction Attn

Filtering the images containing characters which are not in opt.character
Filtering the images whose label is longer than opt.batch_max_length
--------------------------------------------------------------------------------
dataset_root: dataset/train
opt.select_data: ['/']
opt.batch_ratio: ['1']
--------------------------------------------------------------------------------
dataset_root:    dataset/train	 dataset: /
sub-directory:	/.	 num samples: 70
num total samples of /: 70 x 1.0 (total_data_usage_ratio) = 70
num samples of / per batch: 192 x 1.0 (batch_ratio) = 192
/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py:557: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freez

در فایل train.py ⁉
batch_max_length = default=8




num_iter = (300000 / batch size=192 ) / number of batches